cmd/kubeadm/app/phases/upgrade/versiongetter_test.go

```diff
@@ -0,0 +1,306 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"reflect"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/version"
+	fakediscovery "k8s.io/client-go/discovery/fake"
+	clientsetfake "k8s.io/client-go/kubernetes/fake"
+)
+
+func Test_computeKubeletVersions(t *testing.T) {
+	tests := []struct {
+		name  string
+		nodes []v1.Node
+		want  map[string]uint16
+	}{
+		{
+			name: "kubelet version info exist",
+			nodes: []v1.Node{
+				{
+					Status: v1.NodeStatus{
+						NodeInfo: v1.NodeSystemInfo{
+							KubeletVersion: "1.28.1",
+						},
+					},
+				},
+				{
+					Status: v1.NodeStatus{
+						NodeInfo: v1.NodeSystemInfo{
+							KubeletVersion: "1.28.2",
+						},
+					},
+				},
+				{
+					Status: v1.NodeStatus{
+						NodeInfo: v1.NodeSystemInfo{
+							KubeletVersion: "1.28.2",
+						},
+					},
+				},
+			},
+			want: map[string]uint16{
+				"1.28.1": 1,
+				"1.28.2": 2,
+			},
+		},
+		{
+			name: "kubelet version info empty",
+			nodes: []v1.Node{
+				{
+					Status: v1.NodeStatus{},
+				},
+				{
+					Status: v1.NodeStatus{},
+				},
+			},
+			want: map[string]uint16{
+				"": 2,
+			},
+		},
+		{
+			name:  "node is empty",
+			nodes: []v1.Node{},
+			want:  map[string]uint16{},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := computeKubeletVersions(tt.nodes); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestKubeVersionGetter_ClusterVersion(t *testing.T) {
+	tests := []struct {
+		name               string
+		version            *version.Info
+		wantclusterVersion string
+		wantparsedVersion  string
+		wantErr            bool
+	}{
+		{
+			name: "cluster version is valid",
+			version: &version.Info{
+				GitVersion: "1.20.0",
+			},
+			wantclusterVersion: "1.20.0",
+			wantparsedVersion:  "1.20.0",
+			wantErr:            false,
+		},
+		{
+			name:               "cluster version is empty",
+			version:            &version.Info{},
+			wantclusterVersion: "",
+			wantparsedVersion:  "",
+			wantErr:            true,
+		},
+		{
+			name: "cluster version is invalid",
+			version: &version.Info{
+				GitVersion: "invalid-version",
+			},
+			wantclusterVersion: "",
+			wantparsedVersion:  "",
+			wantErr:            true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			client := clientsetfake.NewSimpleClientset()
+			client.Discovery().(*fakediscovery.FakeDiscovery).FakedServerVersion = tt.version
+
+			g := &KubeVersionGetter{
+				client: client,
+			}
+			clusterVersion, parsedVersion, err := g.ClusterVersion()
+
+			if (err != nil) != tt.wantErr {
+				t.Errorf("error = %v, wantErr %v", err, tt.wantErr)
```

Comment:
```suggestion
				t.Errorf("error = %v, wantErr = %v", err, tt.wantErr)
```

more problems like this below...